### ISSUE TITLE
DON'T USE !!! The focal loss is not right.

### ISSUE THREAD

**Issue by BigFishMaster** [ZhaoJ9014/face.evoLVe.PyTorch#120]
The focal loss is not right absolutely. Why nobody point it out ?
https://github.com/ZhaoJ9014/face.evoLVe.PyTorch/blob/d5e31893f7e30c0f82262e701463fd83d9725381/loss/focal.py#L16

**Comment by BigFishMaster** [comment]
@CodePothunter logp will be a scalar for one batch. Actually， it should be a vector for each sample.

**Comment by Samonsix** [comment]
> @CodePothunter logp will be a scalar for one batch. Actually， it should be a vector for each sample.

I agree with you

**Comment by Hzzone** [comment]
You are right, I totally agree with you!

**Comment by Hzzone** [comment]
Maybe a right focal loss:

```python
class FocalLoss(nn.Module):
    def __init__(self, gamma=2):
        super(FocalLoss, self).__init__()
        self.gamma = gamma

    def forward(self, inputs, targets):
        pt = F.softmax(inputs, dim=1)[torch.arange(inputs.size(0)), targets]
        loss = - ((1 - pt) ** self.gamma) * pt.log()
        return loss.mean()

```
